### PR TITLE
core: support multiple jdbc table managers

### PR DESCRIPTION
**Pull request by kokosing**
core: support multiple jdbc table managers

Now user can specify multiple table managers on the same database. This
is particular useful for JdbcTableManger. See tempto-configuration.yaml
in tempto-examples.

ImmutableTablesState had to change, before it had a map table name to table.
Now it has nested map. Map of database name to map of table name to
table. This is because, different databases may have tables with the same
name. The same thing with MutableTablesState.

Ticket: SWARM-884

Test Plan: running tempto examples

Reviewers: losipiuk, ksobczak

### CODE REVIEW

**Comment by losipiuk** [comment]
cc: @brian-rickman

File: tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/ImmutableTablesState.java [hunk 1]
```diff
@@ -21,7 +21,7 @@
 import static com.teradata.tempto.context.ThreadLocalTestContextHolder.testContext;
 
 public class ImmutableTablesState
-        extends NamedObjectsState<TableInstance>
+        extends NamedObjectsState<NamedObjectsState<TableInstance>>
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/ImmutableTablesState.java]: Add helper method which takes both tableManagerName and tableName

File: tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TableManager.java [hunk 2]
```diff
@@ -34,6 +33,7 @@
     @interface Descriptor
     {
         Class<? extends TableDefinition> tableDefinitionClass();
+
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TableManager.java]: why?

File: tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TableManagerHandle.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.teradata.tempto.fulfillment.table;
+
+import static java.util.Objects.requireNonNull;
+
+public class TableManagerHandle
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TableManagerHandle.java]: I think TableManagerKey would be better name for this class

File: tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TableManagerHandle.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    public String getName()
+    {
+        return name;
+    }
+
+    public Class<? extends TableDefinition> getTableDefinitionClass()
+    {
+        return tableDefinitionClass;
+    }
+}
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TableManagerHandle.java]: new line + change intellij settings

File: tempto-core/src/main/java/com/teradata/tempto/internal/convention/sql/SqlPathTestFactory.java [hunk 5]
```diff
@@ -146,14 +146,15 @@ private Requirement getRequirements(SqlQueryDescriptor queryDescriptor)
...
-                .map(requiredTableName -> new ImmutableTableRequirement(tableDefinitionsRepository.getForName(requiredTableName)))
+                .map(requiredTableName -> new ImmutableTableRequirement(tableDefinitionsRepository.getForName(requiredTableName), Optional.of(queryDescriptor.getDatabaseName())))
                 .collect(toList()));
         requirements.addAll(queryDescriptor.getMutableTableDescriptors()
                 .stream()
-                .map(descriptor -> new MutableTableRequirement(
-                        tableDefinitionsRepository.getForName(descriptor.tableDefinitionName),
-                        descriptor.name,
-                        descriptor.state))
+                .map(descriptor -> MutableTableRequirement.builder(tableDefinitionsRepository.getForName(descriptor.tableDefinitionName))
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/convention/sql/SqlPathTestFactory.java]: Maybe replace this .map line with
```
.map(TableDescriptor::getTableDefinitionName)
.map(tableDefinitionRepository::getForName)
.map(MutableTableRequirement::builder)
.map(builder -> ....)
```

Is that more readable?

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/ImmutableTablesFulfiller.java [hunk 6]
```diff
@@ -73,16 +78,45 @@ public void cleanup()
...
     }
 
-    private void createImmutableTable(TableDefinition tableDefinition)
+    private void createImmutableTable(ImmutableTableRequirement tableRequirement)
     {
-        TableManager tableManager = tableManagerDispatcher.getTableManagerFor(tableDefinition);
+        TableDefinition tableDefinition = tableRequirement.getTableDefinition();
+        Optional<String> tableManagerName = tableRequirement.getTableManagerName();
+        TableManagerHandle tableManagerHandle = tableManagerDispatcher.getTableManagerHandleFor(tableDefinition, tableManagerName);
+        TableManager tableManager = tableManagerDispatcher.getTableManagerFor(tableManagerHandle);
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/ImmutableTablesFulfiller.java]: extract obtaining TableManager to separate method

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/TableManagerDispatcherModuleProvider.java [hunk 7]
```diff
@@ -98,16 +104,51 @@ protected void configure()
 
             @Inject
             @Provides
-            public TableManagerDispatcher defaultTableManagerDispatcher(Map<Class, TableManager> tableManagers)
+            public TableManagerDispatcher defaultTableManagerDispatcher(Map<TableManagerHandle, TableManager> tableManagers)
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/TableManagerDispatcherModuleProvider.java]: this is so hardcore. Lets pair program through this.

File: tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/ImmutableTablesState.java (hunk 1 above)
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/ImmutableTablesState.java]: Ignored as I refactored this code much.

File: tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TableManager.java (hunk 2 above)
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TableManager.java]: fixed

File: tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TableManagerHandle.java (hunk 3 above)
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TableManagerHandle.java]: done

File: tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TableManagerHandle.java (hunk 4 above)
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/fulfillment/table/TableManagerHandle.java]: done

File: tempto-core/src/main/java/com/teradata/tempto/internal/convention/sql/SqlPathTestFactory.java (hunk 5 above)
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/internal/convention/sql/SqlPathTestFactory.java]: I cannot do that, sorry.

**Pull request merged**